sdk/schemaregistry/schema-registry-avro/src/schemaRegistryAvroSerializer.ts

```diff
@@ -0,0 +1,226 @@
+    payload.copy(buffer, PAYLOAD_OFFSET);
+    return buffer;
+  }
+
+  // REVIEW: signature. See serialize and s/serialize into/deserialize from/.
+  /**
+   * Deserializes a value from a buffer.
+   *
+   * @param buffer The buffer with the serialized value.
+   * @return The deserialized value.
+   */
+  async deserialize(buffer: Buffer): Promise<any> {
+    if (buffer.length < PAYLOAD_OFFSET) {
+      throw new Error("Buffer is too small to have the correct format.");
+    }
+
+    const format = buffer.readUInt32BE(0);
+    if (format !== FORMAT_INDICATOR) {
+      throw new Error(`Buffer has unknown format indicator: 0x${format.toString(16)}`);
+    }
+
+    const schemaIdBuffer = buffer.slice(SCHEMA_ID_OFFSET, PAYLOAD_OFFSET);
+    const schemaId = schemaIdBuffer.toString("utf-8");
+    const schema = await this.getSchemaById(schemaId);
+    const payloadBuffer = buffer.slice(PAYLOAD_OFFSET);
+
+    return schema.type.fromBuffer(payloadBuffer);
+  }
+
+  //
+  // Avoid hitting the schema registry on every call by caching schemas.
+  //
+  // Currently the cache can only be discarded by throwing away the serializer
+  // and creating a new one, but this will be revisited after the initial
+  // preview:
+  //
+  // https://github.com/Azure/azure-sdk-for-js/issues/10438
+  //
+  private readonly cacheByContent = new Map<string, CacheEntry>();
+  private readonly cacheById = new Map<string, CacheEntry>();
+
+  private async getSchemaById(schemaId: string): Promise<CacheEntry> {
+    const cached = this.cacheById.get(schemaId);
+    if (cached) {
+      return cached;
+    }
+
+    const schemaResponse = await this.registry.getSchemaById(schemaId);
+    if (!schemaResponse.serializationType.match(/^avro$/i)) {
+      throw new Error(
+        `Schema with ID '${schemaResponse.id}' has has serialization type '${schemaResponse.serializationType}', not 'avro'.`
+      );
+    }
+
+    const avroType = avro.Type.forSchema(JSON.parse(schemaResponse.content));
+    return this.cache(schemaId, schemaResponse.content, avroType);
+  }
+
+  private async getSchemaByContent(schema: string): Promise<CacheEntry> {
+    let cached = this.cacheByContent.get(schema);
+    if (cached) {
+      return cached;
+    }
+
+    const avroType = avro.Type.forSchema(JSON.parse(schema));
+    if (!avroType.name) {
+      throw new Error("Schema must have a name.");
+    }
+
+    const description = {
+      group: this.schemaGroup,
+      name: avroType.name,
+      serializationType: "avro",
+      content: schema
+    };
+
+    const schemaIdResponse = this.autoRegisterSchemas
+      ? await this.registry.registerSchema(description)
+      : await this.registry.getSchemaId(description);
+
+    return this.cache(schemaIdResponse.id, schema, avroType);
+  }
+
+  private cache(id: string, schema: string, type: avro.Type): CacheEntry {
+    const entry = { id, type };
+    this.cacheByContent.set(schema, entry);
```

Comment:
huh, I wonder what the perf is for using very large strings as keys....

Reply:
Yes, I wondered about this too. I'll do some measurement. Is there any way in JS to map based on the string instance rather than the string value? I would be fine with a cache miss if the same value came in two different copies of the string in memory.

Reply:
This requires deeper runtime knowledge than I possess. Maybe @bterlson knows how this is implemented?

Reply:
I am not sure how scientific my test is, as I can believe I'm missing a lot of potentially compounding factors, but I am observing ~0.5 to 1 millisecond per megabyte on map.get or set with large key. 

I thought of a way to eliminate this: change the schema arg from string to a new AvroSchema type, which would map by reference in the cache. We'd do the work to parse and create avsc.Type when you created AvroSchema. 

I had thought about taking avsc.Type when I first made this lib, but I am hoping to keep avsc out of the public surface so that we could theoretically change avro implementations in the future. There was some debate about which lib to use in the first place. (This thinking is also where some of my REVIEW questions about minor avsc leakage was coming from.) But if we make our own type that encapsulates avsc.Type, I think we can get the best of both worlds.

I'm thinking this would be a good change, but also that it's large enough that I'd want to get it reviewed independently from this one. I also think we can get away with deferring it to after beta 1. That said, if we really want it or beta 1, I can probably put it together in a couple of hours, but I'd still want it in its own PR at this point.

Thoughts?

Reply:
I agree we can track this as an issue for beta2. Thanks for digging into it a bit!

Reply:
Added to #10950

Reply:
Update here: When I reuse the same string for all calls, get/set time does not vary with string size on average. You take a hit on the first call, and then it's neglible afterwords. It looks like runtime is storing the hashCode with the string on first hash, and reusing it subsequently.

I still think the design change proposed above is better, but I think the perf as is is not too bad either.